chromadb/types.py

```diff
@@ -0,0 +1,122 @@
+from typing import TypeVar, Optional, Union, Sequence, Dict, Any
+from typing_extensions import Literal, Protocol, TypedDict
+from uuid import UUID
+from enum import Enum
+
+Metadata = Dict[str, Union[str, int, float]]
+
+# Namespaced Names are mechanically just strings, but we use this type to indicate that
+# the intent is for the value to be globally unique and semantically meaningful.
+NamespacedName = str
+
+
+class ScalarEncoding(Enum):
+    FLOAT16 = "FLOAT16"
+    FLOAT32 = "FLOAT32"
+    INT32 = "INT32"
+
+
+# Note: This is the data model for identifying and describing an embedding function,
+# not the actual function implementation.
+class EmbeddingFunction(TypedDict):
+    name: NamespacedName
+    dimension: int
+    scalar_encoding: ScalarEncoding
+
+
+class Collection(TypedDict):
+    id: UUID
+    name: str
+    topic: str
+    metadata: Optional[Metadata]
+
+
+class Segment(TypedDict):
+    id: UUID
+    type: NamespacedName
```

Comment:
Could make this a String Enum

Reply:
Decided not to, to avoid having duplicate code.

There will need to be a place in the code where we map Segment Type->Implementation Class so we can instantiate them. If we made this type an enum, we'd need to duplicate that list and have values both in a mapping and in an enum.

I'd rather just have that mapping be the canonically supported set of types,  even if we don't get type system support for that. I don't think there's a way to use the type system to say "this value must be present as a key in this other dict."